Expose DRA device health in PodStatus

Jpsassine opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
This PR implements the DRA portion of KEP-4680, introducing a mechanism for the Kubelet to monitor and report the health of devices allocated via Dynamic Resource Allocation (DRA). This provides cluster operators and controllers with crucial visibility into device health directly in the Pod's status, enabling better troubleshooting and remediation of pod failures caused by faulty hardware.

The core changes are:

1.  **New gRPC Health Service (`dra-health/v1alpha1`):**
    *   Defines a new `NodeHealth` gRPC service with a `WatchResources` server-streaming RPC.
    *   DRA plugins can optionally implement this service to stream device health status updates (`Healthy`, `Unhealthy`, `Unknown`) to the Kubelet.

2.  **Kubelet Integration:**
    *   The `DRAPluginManager` now detects plugins that implement the `NodeHealth` service and starts a long-lived `WatchResources` stream to monitor them. It gracefully handles plugins that do not implement the service.
    *   The **DRA Manager** (`pkg/kubelet/cm/dra`) consumes these health updates and maintains a new, persistent `healthInfoCache` to track the last known state of each device. This cache handles device timeouts and survives Kubelet restarts.
    *   When a device's health changes, the DRA Manager identifies all affected pods and triggers an asynchronous pod sync via the **Container Manager**.

3.  **API Change (`v1.PodStatus`):**
    *   A new field, `allocatedResourcesStatus`, is added to `v1.ContainerStatus`.
    *   During a pod sync, the DRA Manager's `UpdateAllocatedResourcesStatus` function populates this field with the current health of the container's allocated devices, read from the health cache.
    *   The `status_manager`'s equality check (`isPodStatusByKubeletEqual`) is updated to correctly detect changes within this new field, ensuring the pod status is patched to the API server.

4.  **Testing:**
    *   Adds comprehensive unit tests for the new health caching and processing logic.
    *   Adds a new node e2e test (`[Feature:DynamicResourceAllocation] Resource Health`) that verifies the end-to-end flow: a change in device health reported by a test driver is correctly reflected in the `PodStatus` on the API server.
    *   Fixes discovered in existing e2e tests during development are included to ensure the suite remains stable.

#### Which issue(s) this PR fixes:
Fixes #126243

#### Special notes for your reviewer:
This PR focuses on the DRA implementation path for KEP-4680. All changes are behind the `ResourceHealthStatus` feature gate.

#### Does this PR introduce a user-facing change?
Yes. When the `ResourceHealthStatus` feature gate is enabled, a new `allocatedResourcesStatus` field will appear in the `containerStatuses` of a Pod's status if it uses Dynamic Resource Allocation. This allows users to inspect the health of the allocated devices.

```release-note
The Kubelet can now monitor the health of devices allocated via Dynamic Resource Allocation (DRA) and report it in the `pod.status.containerStatuses.allocatedResourcesStatus` field. This requires the DRA plugin to implement the new v1alpha1 `NodeHealth` gRPC service. This feature is controlled by the `ResourceHealthStatus` feature gate.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4680-add-resource-health-to-pod-status/README.md


```

googs1025 commented:
/ok-to-test

bart0sh commented:
@Jpsassine Thank you for your PR. Please [sign the CLA](https://github.com/kubernetes/community/tree/master/contributors/guide/contributor-cheatsheet#signing-the-cla) to proceed further, thanks.

BTW, is there any KEP or another design document describing/discussing these changes? If so, please provide links in the PR description.

Jpsassine commented:
/easycla

bart0sh commented:
/cc

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -580,6 +580,68 @@ func (m *ManagerImpl) GetContainerClaimInfos(pod *v1.Pod, container *v1.Containe
+			}
+
+			// Initialize AllocatedResourcesStatus if nil
+			if status.ContainerStatuses[i].AllocatedResourcesStatus == nil {
```
nit:
```suggestion
			if containerStatus.AllocatedResourcesStatus == nil {
```

ArangoGutierrez commented:
/cc

ArangoGutierrez commented:
/sig-node
/wg device-management

ArangoGutierrez commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
```suggestion
/*
Copyright 2025 The Kubernetes Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0
    
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

ArangoGutierrez replied:
```suggestion
/*
Copyright 2024 The Kubernetes Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

ArangoGutierrez commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,192 @@
+)
+
+const (
+	healthTimeout = 30 * time.Second
```
Is there a reason for this time selection?

Jpsassine replied:
It is currently arbitrary, I would like input on what a good expiration time would be.

harche replied:
Would it make sense to set this value to [RuntimeRequestTimeout](https://github.com/kubernetes/kubernetes/blob/e84a6cb7ef17538ca7c3ee02f1b69f5c7fef7059/pkg/kubelet/apis/config/types.go#L282) instead of an arbitrary value? This value defaults to 2 mins. 

If not, can it be made configurable just like `RuntimeRequestTimeout`?

SergeyKanzhelev replied:
I wonder if we want to make it per-plugin instead? Make it a part of an API between the kubelet and DRA plugin? So some plugins can say "the health doesn't really have to be updated this often" and some will ask for a lot of updates.

@Jpsassine how easy will it be to add this to API? I do not want to block on this and maybe we can do it as a follow up PR

ArangoGutierrez commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,28 @@
+
+message WatchResourcesResponse {
+  repeated DeviceHealth devices = 1;
+}
```
NIT: New line

ArangoGutierrez commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,28 @@
+syntax = "proto3";
```
Is this file autogenerated?  if not, is missing the LICENSE header

ArangoGutierrez commented on pkg/kubelet/cm/dra/plugin/types.go:
```diff
@@ -0,0 +1,32 @@
+/*
+   Copyright 2024 The Kubernetes Authors.
```
```suggestion
   Copyright 2025 The Kubernetes Authors.
```

ArangoGutierrez commented on pkg/kubelet/cm/dra/plugin/types.go:
```diff
@@ -0,0 +1,32 @@
+   limitations under the License.
+*/
+
+// Ensure this is the correct package declaration
```
??

ArangoGutierrez commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -177,11 +183,40 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
 	}
 
+	// --- Start Health Stream Logic ---
+	// Create a separate context derived from the plugin's context for the stream attempt/lifecycle
```
```suggestion
	// Create a separate context derived from the plugin's context for the stream attempt/lifecycle.
```

ArangoGutierrez commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -177,11 +183,40 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
 		clientCallTimeout: timeout,
 	}
 
+	// --- Start Health Stream Logic ---
```
```suggestion
```

ArangoGutierrez commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -89,22 +106,51 @@ func (p *Plugin) getOrCreateGRPCConn() (*grpc.ClientConn, error) {
 
+	logger.V(4).Info("Successfully established gRPC connection", "endpoint", p.endpoint)
 	p.conn = conn
+	// Initialize clients now that connection is established
```
```suggestion
	// Initialize clients now that connection is established.
```

ArangoGutierrez reviewed (changes requested):
First pass

ArangoGutierrez commented:
/kind feature

ArangoGutierrez commented:
/release-note-none

ArangoGutierrez commented on pkg/kubelet/cm/dra/types.go:
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package dra
 
 import (
+	"_/usr/local/google/home/jpsassine/kubernetes/kubernetes/pkg/kubelet/cm/resourceupdates"
```
tests are failing do to this import, you need to add the changes to this branch

Jpsassine committed:
feat(kubelet): Add ResourceHealthStatus for DRA pods

This change introduces the ability for the Kubelet to monitor and report
the health of devices allocated via Dynamic Resource Allocation (DRA).
This addresses a key part of KEP-4680 by providing visibility into
device failures, which helps users and controllers diagnose pod failures.

The implementation includes:
- A new `v1alpha1.NodeHealth` gRPC service with a `WatchResources`
  stream that DRA plugins can optionally implement.
- A health information cache within the Kubelet's DRA manager to track
  the last known health of each device and handle plugin disconnections.
- An asynchronous update mechanism that triggers a pod sync when a
  device's health changes.
- A new `allocatedResourcesStatus` field in `v1.ContainerStatus` to
  expose the device health information to users via the Pod API.

Update vendor

KEP-4680: Fix lint, boilerplate, and codegen issues

Add another e2e test, add TODO for KEP4680 & update test infra helpers

Add Feature Gate e2e test

Fixing presubmits

Fix var names, feature gating, and nits

Fix DRA Health gRPC API according to review feedback

haosdent commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// saveToCheckpoint saves the cache to the state file, acquiring the lock.
+func (cache *healthInfoCache) saveToCheckpoint() error {
```
Seems this method didn't be used in this pull request.

Jpsassine replied:
Ah, thanks! I had to update the code to create a version without the lock "saveToCheckpointInternal" as this was causing infinite go routine issues during unit testing. Will remove the dead code.

bart0sh commented:
@Jpsassine Would it make sense to use [grpc healph checking API](https://grpc.io/docs/guides/health-checking/) instead implementing own gRPC service?

Jpsassine commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -1048,10 +1048,42 @@ func (cm *containerManagerImpl) UpdateAllocatedResourcesStatus(pod *v1.Pod, stat
+	var wg sync.WaitGroup
+
+	sources := []<-chan resourceupdates.Update{}
+	if cm.deviceManager != nil {
```
Good question, I added it for robustness incase this causes a panic in the future where it is conditionally created(eg a migration off device plugin).

Jpsassine commented on pkg/kubelet/cm/dra/healthinfo_test.go:
```diff
@@ -0,0 +1,404 @@
+
+// TestUpdateHealthInfo tests adding, updating, and reconciling device health.
+func TestUpdateHealthInfo(t *testing.T) {
+	tmpFile := "/tmp/health_checkpoint_test"
+	defer os.Remove(tmpFile)
```
Great point, I've updated the logic to reset the in-memory map if the checkpoint file is not found. Thanks for catching that!

Jpsassine commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		if os.IsNotExist(err) {
+
+			return nil
```
Great point, I've updated the logic to reset the in-memory map if the checkpoint file is not found. Thanks for catching that!

Jpsassine commented:
/retest-required

Jpsassine commented:
/label api-review

Jpsassine commented:
/remove-label area/apiserver area/cloudprovider area/code-generation area/dependency area/kube-proxy area/kubectl area/release-eng area/test sig/api-machinery sig/apps sig/auth sig/cli sig/cloud-provider sig/cluster-lifecycle sig/etcd sig/instrumentation sig/network sig/release sig/scheduling sig/storage sig/testing

mrunalp commented:
We need API review from API reviewers for @liggitt :+1:

mrunalp commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+package v1alpha1;
+option go_package = "k8s.io/kubelet/pkg/apis/dra-health/v1alpha1";
+
+service NodeHealth {
```
Should this be called `ResourceHealth` or similar?

Jpsassine replied:
Yes, good catch. This was from earlier work on this PR and stuck around

mrunalp commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -136,6 +136,8 @@ type containerManagerImpl struct {
 	// kubeClient is the interface to the Kubernetes API server. May be nil if the kubelet is running in standalone mode.
 	kubeClient clientset.Interface
+	// updates is a channel that provides resource updates.
+	updates chan resourceupdates.Update
```
```suggestion
	resourceUpdates chan resourceupdates.Update
```

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(cache.stateFile, data, 0644)
```
Can we write this atomically?

Jpsassine replied:
yes we can, updated this.

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+	return json.Unmarshal(data, cache.HealthInfo)
+}
+
+// withLock runs a function while holding the claimInfoCache lock.
```
```suggestion
// withLock runs a function while holding the healthInfoCache lock.
```

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+	return f()
+}
+
+// withRLock runs a function while holding the claimInfoCache rlock.
```
```suggestion
// withRLock runs a function while holding the healthInfoCache rlock.
```

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+// getHealthInfo returns the current health info, adjusting for timeouts.
+func (cache *healthInfoCache) getHealthInfo(driverName, poolName, deviceName string) state.DeviceHealthString {
+	res := state.DeviceHealthStringUnknown
+	now := time.Now()
```
Should this be moved inside the lock section below?

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+	_ = cache.withRLock(func() error {
+		if driver, ok := (*cache.HealthInfo)[driverName]; ok {
+			for _, device := range driver.Devices {
+				if device.PoolName == poolName && device.DeviceName == deviceName {
```
Can we use a map for lookup here instead of iterating over a list?

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+
+		if changed {
+			if err := cache.saveToCheckpointInternal(); err != nil {
+				klog.Background().Error(err, "Failed to save health checkpoint after update")
```
Why is err not propagated for this case?

SergeyKanzhelev replied:
nit: update log message to something that will explain what it means:

```suggestion
				klog.Background().Error(err, "Failed to save health checkpoint after update. Kubelet restart may lose the device health information.")
```

mrunalp commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -663,3 +692,209 @@ func (m *Manager) GetContainerClaimInfos(pod *v1.Pod, container *v1.Container) (
+				for driverName, driverState := range claimInfo.DriverState {
+					// Iterate through each specific device allocated by this driver
+					for _, device := range driverState.Devices {
+						m.healthInfoMutex.Lock()
```
healthInfoCache already has a mutex, right? why do we need another one here?

kannon92 commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -350,6 +352,37 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+	if cm.deviceManager != nil {
+		sources["deviceManager"] = cm.deviceManager.Updates()
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DynamicResourceAllocation) && cm.draManager != nil {
```
I think we would need a Device Health feature gate in addition to DRA?

DRA will be going stable this release so this would always be true.

SergeyKanzhelev replied:
+1, should be protected by KEP's FG

kannon92 commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -1050,10 +1083,12 @@ func (cm *containerManagerImpl) UpdateAllocatedResourcesStatus(pod *v1.Pod, stat
 
-	// TODO(SergeyKanzhelev, https://kep.k8s.io/4680): add support for DRA resources which is planned for the next iteration of a KEP.
+	// Update DRA resources if the feature is enabled and the manager exists
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DynamicResourceAllocation) && cm.draManager != nil {
```
same here: https://github.com/kubernetes/kubernetes/pull/130606/files#r2217857219

kannon92 commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -223,7 +223,21 @@ func isPodStatusByKubeletEqual(oldStatus, status *v1.PodStatus) bool {
 		}
 	}
 
+	if !apiequality.Semantic.DeepEqual(oldStatus.ContainerStatuses, status.ContainerStatuses) {
```
I don't understand how this is related to this PR.

Jpsassine replied:
I can remove it, I had added it to reduce flakiness during debugging e2e test issues I had seen. I was seeing errors occur in this function where the container status slice would have an updated AllocatedResourcesStatus field but the DeepEqual would not catch it and therefore the update would not propagate to the API server. 

After adding this more robust, explicit check the test began passing. That being said, I tried reverting this code and running my e2e node tests again and I am not seeing any flakes. Perhaps the issue was coming from elsewhere. I could revert the changes here if bolstering this function is more redundant than useful.

SergeyKanzhelev replied:
let's revert this change if we do not know why it may not work

kannon92 commented on test/e2e_node/dra_test.go:
```diff
@@ -820,10 +820,163 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
 		)
 	})
+
+	f.Context("Resource Health [FeatureGate:ResourceHealthStatus]", f.WithSerial(), func() {
```
Test failures are saying that this is deprecated. You should use `WithFeatureGate`.

Jpsassine replied:
My e2e tests work locally as I am passing them the new feature gate flag but I need to update the presubmits I believe to get the CI working with this PR.

SergeyKanzhelev commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,195 @@
+		stateFile:  stateFile,
+	}
+	if err := cache.loadFromCheckpoint(); err != nil {
+		klog.Background().V(3).Info("Failed to load health checkpoint, proceeding with empty cache", "err", err)
```
V3 seems to be too low. Wouldn't even be a default in many cases. Perhaps just info is OK

```suggestion
		klog.Background().Info("Failed to load health checkpoint, proceeding with empty cache", "err", err)
```

SergeyKanzhelev reviewed (commented):
cannot find the logic when DRA plugin disappearing and we need to mark everythind as "unknown" after some timeout. Is it handled and I just missed it?

liggitt commented:
> We need API review from API reviewers for @liggitt 👍

I added it to my backlog but it's behind several other reviews

SergeyKanzhelev commented:
> > We need API review from API reviewers for @liggitt 👍
> 
> I added it to my backlog but it's behind several other reviews

We actually do not update any APIs. We introduce a new device health gRPC, but this is not under API approvership, similar to dra gRPC. Maybe we should adjust ownership, @mrunalp do you think we need to adjust ownership of those proto definitions?

liggitt commented:
> We introduce a new device health gRPC, but this is not under API approvership, similar to dra gRPC

backend grpc APIs need API review since they're just as much of an API in terms of compatibility implications, etc. If there are DRA grpc APIs in packages not flagged for API review, please adjust that

Jpsassine commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+}
+
+message DeviceHealth {
+  // resource_name is the full system-wide identifier: "<driver name>/<pool name>/<device name>".
```
The plugin should provide the driver name, I think having this resource name as the system wide identifier provides robustness. There might be a scenario where two different drivers have the same pool name and device name.

Jpsassine commented:
> cannot find the logic when DRA plugin disappearing and we need to mark everythind as "unknown" after some timeout. Is it handled and I just missed it?

In the case that there is a gRPC stream break, which would be the case if the DRA plugin disappeared, there is a defer func in HandleWatchResourcesStream in manager.go which wipes that driver's health cache info via `m.healthInfoCache.clearDriver(pluginName)`. After this, if a query for any of those devices happen the health will just return as "Unknown". 

If the plugin for some reason were to just stop sending health updates for all its devices because of some internal problem, the health cache will handle this using the health timeout and the next time the health cache is queried for specific devices they will be marked as "Unknown" after surpassing this health timeout.

Jpsassine commented on test/e2e_node/dra_test.go:
```diff
@@ -820,10 +820,163 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
 		)
 	})
+
+	f.Context("Resource Health [FeatureGate:ResourceHealthStatus]", f.WithSerial(), func() {
```
My e2e tests work locally as I am passing them the new feature gate flag but I need to update the presubmits I believe to get the CI working with this PR.

Jpsassine commented:
I've made a PR to exclude ResourceHealthStatus feature labeled DRA tests from the DRA presubmit job. Currently, my last failing presubmit is `pull-kubernetes-node-e2e-crio-cgrpv1-dra` because 2/3 new e2e tests require the `ResourceHealthStatus` flag to be enabled along with the DRA feature flag. Instead of enabling this new flag for all DRA tests I will use the [existing presubmit job](https://github.com/kubernetes/test-infra/pull/33059/files) for `ResourceHealthStatus` feature tests by manually adding it to the PR. 

Here is my PR for updating the dra.jinja file in `test-infra` https://github.com/kubernetes/test-infra/pull/35195

/test pull-kubernetes-node-e2e-resource-health-status

cc. @SergeyKanzhelev @kannon92

Jpsassine commented:
/test pull-kubernetes-node-e2e-resource-health-status

SergeyKanzhelev commented:
> /test pull-kubernetes-node-e2e-resource-health-status

This run doesn't run any DRA tests: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130606/pull-kubernetes-node-e2e-resource-health-status/1948127574433468416

You may need to do some adjustments in test definition

SergeyKanzhelev commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -595,6 +613,12 @@ func (m *Manager) unprepareResources(ctx context.Context, podUID types.UID, name
 
 	// Atomically perform some operations on the claimInfo cache.
 	err := m.cache.withLock(func() error {
+		// TODO(#132978): Re-evaluate this logic to support post-mortem health updates.
+		// As of the initial implementation, we immediately delete the claim info upon
+		// unprepare. This means a late-arriving health update for a terminated pod
+		// will be missed. A future enhancement could be to "tombstone" this entry for
+		// a grace period instead of deleting it.
```
side note: I wonder if the solution here may be covering both device plugin and DRA

richabanker commented:
/remove-sig instrumentation

msau42 commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+package v1alpha1;
+option go_package = "k8s.io/kubelet/pkg/apis/dra-health/v1alpha1";
+
+service ResourceHealth {
```
service name should be prepended with DRA

SergeyKanzhelev replied:
sure, it will help consistency

msau42 commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+option go_package = "k8s.io/kubelet/pkg/apis/dra-health/v1alpha1";
+
+service ResourceHealth {
+  rpc WatchResources(WatchResourcesRequest) returns (stream WatchResourcesResponse) {}
```
Does this run on nodes? Should we preprend `Node` to the method name?

SergeyKanzhelev replied:
is it for alignement with other DRA methods? This is not the general rule ([example](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1/api.proto)), but I see your point, makes sense

msau42 commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+  // health is the device's status: "Healthy", "Unhealthy", "Unknown".
+  string health = 4;
+  // last_updated is the Unix timestamp (seconds) of the last health update.
+  int64 last_updated = 5;
```
From https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/dra/v1/api.proto, the field name should be suffixed with `_time`.

msau42 commented:
I'm doing an API review on the proto.

SergeyKanzhelev commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+  // This is the canonical, unique ID for the device.
+  string resource_name = 1;
+  // pool_name identifies the pool within the driver.
+  string pool_name = 2;
```
Do you think we need to import other proto or just redefine the same type?

msau42 commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,42 @@
+  // device_name identifies the device within the pool.
+  string device_name = 3;
+  // health is the device's status: "Healthy", "Unhealthy", "Unknown".
+  string health = 4;
```
Do you want all drivers to report a consistent set of strings? If so, then maybe an enum is better.

Jpsassine replied:
Yes, I have changed it to Enum. I am testing my changes and should have them out shortly.

msau42 commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,67 @@
+  DeviceIdentifier device = 1;
+
+  // The health status of the device.
+  HealthStatus health = 2; // Assuming you are also making the enum change
```
Would you want a way for drivers to return a more detailed string/msg? Similar to how we do conditions in k8s today.

Jpsassine replied:
I think this is a good idea, but I think this would be part of the future enhancements to the KEP that we introduce when the feature goes beta in 1.35.

Jpsassine commented on staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto:
```diff
@@ -0,0 +1,67 @@
+}
+
+// HealthStatus defines the possible health states of a device.
+enum HealthStatus {
```
Yes, we do plan on potentially adding more states in the future. The Kubelet's implementation will handle this by treating any unrecognized enum value as `UNKNOWN`, ensuring forward compatibility as you suggested.

msau42 commented:
Approving the alpha proto API changes as API reviewer.

I did not review anything else.

Jpsassine commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	defer func() {
+		if err := os.Remove(tempFile.Name()); err != nil && !os.IsNotExist(err) {
```
this is to cleanup the temp file incase an error occurs during the middle of the write process

mrunalp replied:
We can attempt to cleanup if the rename (move) failed but if it succeeded there should be no file present in the original location.

Jpsassine replied:
Yes but this is for a case in which either tempFile.Write or tempFile.Close were to fail. In this failure case we would exit the function but the tempFile would be stuck with no cleanup.

mrunalp commented on pkg/kubelet/cm/dra/healthinfo.go:
```diff
@@ -0,0 +1,214 @@
+			if _, wasReported := reportedKeys[key]; !wasReported {
+
+				if existingDevice.Health != state.DeviceHealthStatusUnknown && now.Sub(existingDevice.LastUpdated) > healthTimeout {
+
```
nit: remove empty line

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

chatted with @mrunalp we are ok on sig node.

SergeyKanzhelev commented:
@dims are you around to approve for `hack/`?

SergeyKanzhelev commented:
@dchen1107 are you OK to approve for DRA? @pohly and @klueska are likely will not be around till code freeze

dims commented:
/approve

(for `/hack`)

dchen1107 commented:
/approve to unblock the progress

dims commented:
thanks @dchen1107 

/hold cancel

pacoxu commented:
/test pull-kubernetes-e2e-kind-ipv6

pohly commented on test/e2e_node/dra_test.go:
These tests are failing or flaking:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133214/pull-kubernetes-node-e2e-crio-cgrpv1-dra/1948697267229691904

pohly commented:
https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=130606 shows that the new e2e_node test didn't work.
